src/doc/en/developer/coding_basics.rst

```diff
@@ -153,17 +153,57 @@ Adding new top-level packages below :mod:`sage` should be done
 sparingly.  It is often better to create subpackages of existing
 packages.
 
-Non-Python Sage source code and supporting files can be included in one
-of the following places:
+Non-Python Sage source code and small supporting files can be
+included in one of the following places:
 
 - In the directory of the Python code that uses that file.  When the
   Sage library is installed, the file will be installed in the same
-  location as the Python code. For example,
-  ``SAGE_ROOT/src/sage/interfaces/maxima.py`` needs to use the file
-  ``SAGE_ROOT/src/sage/interfaces/maxima.lisp`` at runtime, so it refers
-  to it as ::
+  location as the Python code. This is referred to as "package data".
+
+  The preferred way to access the data from Python is using the
+  `importlib.resources API <https://importlib-resources.readthedocs.io/en/latest/using.html>`,
+  in particular the function :func:`importlib.resources.files`.
+  It should be imported as follows
+  (see :ref:`section-python-language-standard`)::
+
+    try:
+        # Use backport package providing Python 3.11 features
+        from importlib_resources import files
+    except ImportError:
+        from importlib.resources import files
+
+  After this import, you can:
+
+  - open a resource for text reading: `fd = files(package).joinpath(resource).open('rt')`
+  - open a resource for binary reading: `fd = files(package).joinpath(resource).open('rb')`
+  - read a resource as text: `text = files(package).joinpath(resource).read_text()`
+  - read a resource as bytes: `bytes = files(package).joinpath(resource).read_bytes()`
+  - open a xz resource for text reading: `fd = lzma.open(files(package).joinpath(resource), 'rt')`
+  - open a xz resource for binary reading: `fd = lzma.open(files(package).joinpath(resource), 'rb')`
```

Comment:
```suggestion
  - open a xz resource for text reading: `fd = lzma.open(files(package).joinpath(resource).open('rb'), 'rt')`
  - open a xz resource for binary reading: `fd = lzma.open(files(package).joinpath(resource).open('rb'), 'rb')`
```

It seems `lzma.open()` accepts either a filename or an open stream, so the previous suggestion _will not work_ e.g. if the module is zipped.

The two ways to work around this are:
1. use `open('rb')` in the resource (as here, as in matroid-database)
2. use `as_file(...)` to create an actual file to pass the filename to `lzma.open(...)` (as done in conway-polynomials)

I think the first one is simpler and possibly cheaper for the case of module-within-a-zip-file since it doesn't need to extract the xz file from the zip file to the filesystem before decompressing the xz file.

I guess both should satisfy the "type-checker" but what can you expect if you add a trunk to a duck...

The documentation of `lzma.open()` claims that passing `'rt'` results in a `TextIOWrapper` and this matches my experiment.

Reply:
I removed the `.open('rb')` yesterday: https://github.com/gmou3/matroid-database/blob/39b05995006d4a9ce5d07b15d3ff3e66a62a0175/src/matroid_database/__init__.py#L88

How can I replicate the issue you are mentioning?

Reply:
To reproduce create a zip file with your package (e.g. build the wheel) and run with `PYTHONPATH=path/to/zipfile`. For instance:

```
$ PYTHONPATH=src python
Python 3.12.2 (main, Feb 15 2024, 06:47:30) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from matroid_database import *
>>> for m in all_matroids_revlex(2, 1): print(m)
... 
**
0*
```
works, but
```
$ PYTHONPATH=dist/matroid_database-0.2-py3-none-any.whl python
Python 3.12.2 (main, Feb 15 2024, 06:47:30) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from matroid_database import *
>>> for m in all_matroids_revlex(2, 1): print(m)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tornaria/src/sage/matroid-database/dist/matroid_database-0.2-py3-none-any.whl/matroid_database/__init__.py", line 103, in all_matroids_revlex
ImportError: cannot import name '_all' from 'matroid_database' (/home/tornaria/src/sage/matroid-database/dist/matroid_database-0.2-py3-none-any.whl/matroid_database/__init__.py)
```
doesn't.

However, the error above is not because of `lzma.open(...)`: in a surprising turn of events, it seems zip files don't support namespace packages either so `from . import _all` fails :rofl:

Also `.with_suffix()` doesn't work on resources from zip files. With the following change:

```diff
--- a/src/matroid_database/__init__.py
+++ b/src/matroid_database/__init__.py
@@ -83,7 +83,7 @@ def _open_data(module, name):
     import lzma
     from importlib.resources import files
 
-    path = files(module).joinpath(name).with_suffix('.txt.xz')
+    path = files(module).joinpath(name + '.txt.xz')
     try:
         return lzma.open(path, 'rt')
     except FileNotFoundError:
@@ -100,8 +100,7 @@ def all_matroids_revlex(n, r):
     Return an iterator over the revlex encodings of all matroids of given
     number of elements and rank.
     """
-    from . import _all
-    with _open_data(_all, f"allr{r}n{n:02d}") as f:
+    with _open_data(__package__, f"_all/allr{r}n{n:02d}") as f:
         while s := f.readline():
             yield s.strip()
 
```

Now we get the issue with `lzma.open(..)` not supporting resources in zip files:
```
$ PYTHONPATH=dist/matroid_database-0.2-py3-none-any.whl python -c $'from matroid_database import *\nfor m in all_matroids_revlex(2, 1): print(m)'
Traceback (most recent call last):
  File "<string>", line 2, in <module>
  File "/home/tornaria/src/sage/matroid-database/dist/matroid_database-0.2-py3-none-any.whl/matroid_database/__init__.py", line 103, in all_matroids_revlex
  File "/home/tornaria/src/sage/matroid-database/dist/matroid_database-0.2-py3-none-any.whl/matroid_database/__init__.py", line 88, in _open_data
  File "/usr/lib/python3.12/lzma.py", line 309, in open
    binary_file = LZMAFile(filename, lz_mode, format=format, check=check,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/lzma.py", line 127, in __init__
    raise TypeError("filename must be a str, bytes, file or PathLike object")
TypeError: filename must be a str, bytes, file or PathLike object
```

Now apply:
```diff
--- a/src/matroid_database/__init__.py
+++ b/src/matroid_database/__init__.py
@@ -85,7 +85,7 @@ def _open_data(module, name):
 
     path = files(module).joinpath(name + '.txt.xz')
     try:
-        return lzma.open(path, 'rt')
+        return lzma.open(path.open('rb'), 'rt')
     except FileNotFoundError:
         raise ValueError(
             "unable to open %s" % path +
```

And voila:
```
$ PYTHONPATH=dist/matroid_database-0.2-py3-none-any.whl python -c $'from matroid_database import *\nfor m in all_matroids_revlex(2, 1): print(m)'
**
0*
```


I'm not sure what is the value of supporting running from zip files, but the whole push against `ext_data` and in favor of `importlib.resources` is to support it...

Reply:
BTW, to build or rebuild the wheel I just do
```
$ python -m build -nxw
...
Successfully built matroid_database-0.2-py3-none-any.whl
```
Redo this step each time you change the source.

Reply:
Ok, thanks. I'll do it like this since it is more robust. One last aesthetic thing: is there a reason to call the data folders `_all` and `_unorientable` instead of `all` and `unorientable`?